### ISSUE TITLE
Use sailfishsecretsd as a GPG agent, is it a good idea?

### ISSUE THREAD

**Issue by dcaliste** [sailfishos/sailfish-secrets#5]
From what I've understand of the README, this daemon will provide a secured and unified way to access crypto operations. That's great. There is already a (restricted) framework, but widely used, doing that namely GPG. So I'm brainstorming on how to integrate the two projects.

For reference here is how GPG operations are handled:

* there is a daemon-like process called GPG agent running (or started on demand). This daemon is playing an intermediate role between the application and possible user interaction.
* when a program is doing a crypto operation (like sign for instance), using either gpg command-line program or gpgme API, a IPC connection is opened with the GPG agent if some user interaction is required.
* then, the GPG agent may start a pinentry program and ask it via another IPC the passphrase. Then the passphrase is returned to the application via IPC.
* finally gpgme internal handle the IPC with the GPG program actually doing the operation and returned data to the application through IPC.

Using the default GPG agent is not convenient in Sailfish for these reasons:
* it is starting a new process (the pinentry) each time a user interaction is required.
* it seems to me that it's difficult to integrate this new process into lipstick: how to tell lipstick that this process would have no window, should be modal and displayed as a system dialog?

So my question, why not make the brand new sailfishsecretsd implements a GPG agent? I've done this already as a test for GPG signature in mail, it's not complicated. How would this be done? Which plugin API should be used?

**Comment by chriadam** [comment]
First, yes you're correct about what sailfish-secrets provides.  I agree that it is a good idea to ensure that sailfish-secrets fulfills the needs of all clients, including doing what GPG does currently.

In regards to the user-interaction flows, this is a core consideration for our platform.  Specifically, we are currently adding support to devicelock daemon to allow clients to request devicelock to perform a "generic" user interaction flow (where the actual UI will be shown by lipstick).  In addition to those "system-mediated" UI flows, the secrets API also supports "in-app" flows (in fact, those are already implemented and can be tested right now).

Finally: I'm not sure whether it makes sense to make sailfishsecretsd offer the _same_ API as GPG agent, although it certainly makes sense to support all of the required operations.  Any input / discussion / thoughts about this topic would be appreciated.  Whatever style the API takes, however, the actual implementation will come from (a variety?) of plugins, e.g. EncryptionPlugin-derived type for the PKDECRYPT, PKSIGN operations, StoragePlugin-derived type for the LEARN operation, CryptoPlugin-derived type for the GENKEY, IMPORT, EXPORT, HAVEKEY and ISTRUSTED operations.

I guess we need to determine: what is currently missing from the Sailfish::Secrets/Crypto APIs which need to be added to support all of the GPG operations; how we can implement the required functionality; and how best to expose that for GPG clients.

**Comment by dcaliste** [comment]
> Specifically, we are currently adding support to devicelock daemon to allow clients to request devicelock to perform a "generic" user interaction flow

Reading this and the later part, I may have a better proposition for GPG integration: instead of rewriting a GPG-agent, let's implement a pinentry program that is a DBus program poking sailfishsecretd. Like that the workflow a GPG-only aware program would be:
* a user application is calling ```gpgme_sign_async(priv_key, data)``` for instance and wait for completion of the action. This action requires to "unlock" the private key ```priv_key``` with the user passphrase (or the cached version).
* gpgme library is then contacting (or launching on-demand) the standard GPG-agent from official gpg RPM package.
* this GPG-agent is configured to run ```/usr/bin/pinentry``` (default) which is a new program to write, and ask it the secret passphrase.
* this new ```pinentry```, ask the secret through DBus from sailfishsecretd and return it to the GPG-agent, which in turn returns to the user application, signaling that passphrase is available and operation can be done.

Do you agree with this workfolow ? It allows to write minimal code (only the pinentry which is already desktop dependent anyway, GPG providing one for GTK, Qt or Gnome implementing one itself in Gnome-shell) to interface properly with existing GPG workflow and all non-sailfishsecretd-aware programs should work out-of-the-box.

Can you orientate me which sailfishsecretd API should be used to do that?

> what is currently missing from the Sailfish::Secrets/Crypto APIs which need to be added to support all of the GPG operations

The crypto API from the cryptoplugin looks alright from first glance I gave. I think, I can write a simple plugin allowing crypto operations with the GPG backend easily (it is more or less the same API than gpgme). I should do the job to be 100% sure though. My question here is simple: since gpgme already exists and non Sailfish programs are already using it, what would be the aim of a new program to use the sailfishsecret API instead of the gpgme directly?

**Comment by Venemo** [comment]
@dcaliste Maybe I'm misunderstanding your intentions, but here are my current thoughts. Why not just write a GPG `EncryptionPlugin` for sailfish-secrets? Then there would be only one daemon running (sailfishsecretsd) and you don't need to worry about anything else.

If you want to run code which uses gpgme:  
Perhaps a wrapper could be written which translates the gpgme API calls to sailfishsecretsd API calls. I'm unfamiliar with gpgme so not sure how feasible this is.

**Comment by dcaliste** [comment]
> Why not just write a GPG EncryptionPlugin for sailfish-secrets

That's my plans indeed. I should do it in April guess.

> Then there would be only one daemon running

Yes, I agree for all codes using secrets API. That's the best way to go. But for other codes (mainly coming from Desktop) they will rely on the GPG agent working on device. So my point is to develop a pinentry based on secrets for Sailfish, as Gnome, Plasma, … have develop their own pinentry to interface with GnuPG in an integrated way.

I've done the job, see my pinentry branch.

**Comment by Venemo** [comment]
@dcaliste Thanks! I'll take a look when I finish what I started with the sign / verify stuff. So probably early next week.

**Comment by dcaliste** [comment]
On last note, the title of the issue is now misleading. I was thinking first to implement a full GPG agent with secret API, but that's useless. Simply the pinentry part is required and the official GnuPG agent will do the job as agent.

Years before, I started to think that we needed to reimplement a full GPG agent because the official one is forking to run a pinentry and I thought at that time that it was not convenient because Lipstick will finally handle the UI part and can obviously not by restart at each gpg demand (!!!). So I thought that we needed an agent that can talk to an already running pinentry (aka. lipstick in our case). But I've now evolved and see that the pinentry can be a thin layer that can be run on-demand by the agent and talk to the UI via DBus calls, or even better now via the secret API (which ultimately interface with the UI via DBus).

That's what is done in my pinentry branch. I'm now waiting for the part linking Lipstick and secret together, so when a password is required, a system window like the one for wifi is shown.
